feat: Allowlist

MarcoPolo opened the pull request:
This adds the Allowlist to Resource manager. See docs/allowlist.md for the design doc and more details.

This change requires changes to be merged first in go-libp2p-core and go-multiaddr (to support the ipcidr protocol in the multiaddr).

After this is merged, go-libp2p will need to be plumbed with these changes as well.

Before Merge:
- [x] Update resource manager interface in go-libp2p-core
  - [x] https://github.com/libp2p/go-libp2p-core/pull/257
- [x] Update go-multiaddr + multiaddr repo with `ipcidr`
  - [x] https://github.com/multiformats/go-multiaddr/pull/177
  - [x] https://github.com/multiformats/multiaddr/pull/129

After Merge:
- [ ] Another PR for the config input to define an allowlist. The allowlist supports dynamic updating, but it's not exposed yet.
- [ ] Update go-libp2p to pass in the endpoint information

MarcoPolo commented:
@marten-seemann / @vyzo this is ready to review. Please take a look.

Before merging I'll make sure to finish this checkboxed things in the PR description.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+	}
+}
+
+func toIPNet(ma multiaddr.Multiaddr) (*net.IPNet, string, error) {
```
If you move peer ID decoding to this function, you don't have to do it twice below (in Add and in Delete).

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+
+}
+
+func (al *allowlist) Add(ma multiaddr.Multiaddr) error {
```
Can you add a comment here which kind of multiaddrs are valid (namely, only resolved multiaddrs with and without /p2p)?

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+		return false
+	}
+
+	_ = ip
```
Nit: leftover from developing?

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+
+	_ = ip
+	for _, network := range al.allowedNetworks {
+		_ = network
```
and here

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+	allowedNetworks []*net.IPNet
+
+	// Only the specified peers can use these IPs
+	allowedPeerByNetwork map[peer.ID][]*net.IPNet
```
Have you considered unifying these into a single map? `map[net.IPNet]struct{ allowsAll bool; map[peer.ID]struct{} }`
`allowsAll` would be set if a multiaddr without /p2p is added.

MarcoPolo replied:
Can I use net.IPNet as a key? The struct holds two byteslices, so I don't think I can.

marten-seemann replied:
Just checked, you can't. That's annoying. You _could_ use `net.IPNet.String()`, but not sure if that's nicer.

MarcoPolo replied:
I don’t think so because:

1. allow checking is harder. you have to convert between the string and bytes.
2. You need to allocate the string on insertion. 
3. you need to allocate at least once when doing an allowlist check. 
4. (Depending on check implementation) you lose the quick lookup by peerid. Which may be a pretty common use case.

marten-seemann commented on rcmgr.go:
```diff
@@ -89,10 +95,12 @@ var _ network.PeerScope = (*peerScope)(nil)
-	peer  *peerScope
+	dir           network.Direction
+	usefd         bool
+	isAllowlisted bool
```
Nit: `isAllowListed`

MarcoPolo replied:
I meant to call this `allowlist` or `Allowlist`. Purposely not capitalizing the l. I realize I should change `newAllowList` to `newAllowlist` to be consistent.

marten-seemann commented on rcmgr_test.go:
```diff
@@ -6,8 +6,11 @@ import (
+	"github.com/multiformats/go-multiaddr"
 )
 
+var dummyMA, _ = multiaddr.NewMultiaddr("/ip4/1.2.3.4/tcp/1234")
```
Nit: `ma.StringCast` is nicer in tests.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,190 @@
+			al.allowedPeerByNetwork[allowedPeer] = append(al.allowedPeerByNetwork[allowedPeer], ipnet)
+		}
+	} else {
+		if ipnet != nil {
```
This can never happen, can it?

MarcoPolo replied:
Correct, remnant from a refactor. Thank you

marten-seemann reviewed (commented):
Is my understanding correct, that we allow two types of addresses to be whitelisted?
1. `/ip{4,6}/<ip>/`
2. `/ip{4,6}/<ip>/p2p/<peerid>`
But not `/p2p/<peerid>`?

We also need to add configurable limits for the newly introduced allowlist scopes, right?

marten-seemann commented:
Requesting a review from @vyzo. Can you check how the scopes are getting shifted around?

MarcoPolo commented:
> We also need to add configurable limits for the newly introduced allowlist scopes, right?

Yep! Thank you for reminding me. I just added those with the same limits as the standard scopes.

> Is my understanding correct, that we allow two types of addresses to be whitelisted?
> 
>     1. `/ip{4,6}/<ip>/`
> 
>     2. `/ip{4,6}/<ip>/p2p/<peerid>`
>        But not `/p2p/<peerid>`?

Correct. I'm not sure what value the `/p2p/<peerid>` would give you since we'd have no way of knowing a connection was for this peerid until after we get the peerid information. Maybe if we knew that this address was signed by the `<peerid>` then we could allowlist the connection attempt. But I think you can do that with this by `Add`ing the addresses you discover are signed by the peer you trust to the allowlist with the peer's peer id appended.

marten-seemann commented:
> Correct. I'm not sure what value the /p2p/<peerid> would give you since we'd have no way of knowing a connection was for this peerid until after we get the peerid information.

I agree. I was misled by the `ipNet != nil` check into assuming that we can have addresses that don't have an IP address.

MarcoPolo commented:
Friendly ping @vyzo :)

marten-seemann commented:
@MarcoPolo Do you want to include the config changes in this PR, or are you planning to create a follow-up PR for that?

MarcoPolo commented:
> @MarcoPolo Do you want to include the config changes in this PR, or are you planning to create a follow-up PR for that?

My plan was after this (see pr description). Since this will work fine without it and the config change is pretty small. And you have some changes around config in #48 in progress.

vyzo commented on allowlist.go:
```diff
@@ -0,0 +1,187 @@
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+type allowlist struct {
```
I think we need to make a public interface for it in extapi, and provide an accessor so that users can modify after initialization.
The RWMutex is a must in this case.

MarcoPolo replied:
I don't follow the reason for a public interface? I thought generally interfaces belonged to the caller.

> Interfaces declare the behaviour the caller requires not the behaviour the type will provide. Let callers define an interface that describes the behaviour they expect. The interface belongs to them, the consumer, not you.

From: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_let_callers_define_the_interface_they_require

MarcoPolo replied:
ah is it because we return an interface and we can't have an interface function return this unexported type?

What's the downside of exporting this type then?

vyzo commented on rcmgr.go:
```diff
@@ -500,13 +533,78 @@ func (s *connectionScope) PeerScope() network.PeerScope {
+		transient = s.rcmgr.transient
+	}
+
+	if s.isAllowlisted && !s.rcmgr.allowlist.AllowedPeerAndMultiaddr(p, s.endpoint) {
```
fuse with the if above?

vyzo reviewed (commented):
We need a public interface and accessor in the rcmgr for the allowlist, plus an allowlist option to initialize.

MarcoPolo commented:
> We need a public interface and accessor in the rcmgr for the allowlist, plus an allowlist option to initialize.

I exported the Allowlist type, added an accessor to rcmgr, and added the `WithAllowlistedMultiaddrs` option.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+	var allowedPeer peer.ID
+	var isIPV4 bool
+
+	multiaddr.ForEach(ma, func(c multiaddr.Component) bool {
```
Can we use `ma.MultiaddrToIPNet` here?

marten-seemann replied:
Probably not, if we want to extract the `p2p` component at the same time, right? We could get that one without iterating again by splitting off the last component though.

MarcoPolo replied:
I also need the isIPV4 to save some time below

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+	}
+	ipNetList := al.allowedNetworks
+
+	if allowedPeer != peer.ID("") {
```
```suggestion
	if allowedPeer != "" {
```

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+func (al *Allowlist) Add(ma multiaddr.Multiaddr) error {
+	al.mu.Lock()
+	defer al.mu.Unlock()
+	ipnet, allowedPeer, err := toIPNet(ma)
```
Nit: move the mutex locking after this.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+	al.mu.Lock()
+	defer al.mu.Unlock()
+
+	ipnet, allowedPeer, err := toIPNet(ma)
```
Here too.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	i := len(ipNetList)
+	for i > 0 {
```
Any reason not to use `for i, ipnet := range ipNetList`? Do we gain anything from going through the list in reverse order?

MarcoPolo replied:
For swap remove. If we go through in normal order then we would skip an entry after swap remove or do some other messy logic that I think boils down to reverse iterating

marten-seemann replied:
That only applies if you're trying to make multiple deletions from the list. Should be fine if you're `break`ing after the first deletion, right?

MarcoPolo replied:
ah that's true. Although I'm a bit hesitant to change this so that instead of working for N removals it only works for 1 removal and would subtly break if you try to remove more than 1.

And the only reason to make the change is so that we use the syntactic sugar of `for ... range`. Is there any other benefit?

marten-seemann replied:
Fair enough. I'd consider `range` more idiomatic (makes it harder to produce out-of-bounds access). No need to change it here.

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+		if ipNetList[i].IP.Equal(ipnet.IP) && bytes.Equal(ipNetList[i].Mask, ipnet.Mask) {
+			if i == len(ipNetList)-1 {
+				// Trim this element from the end
+				ipNetList = ipNetList[:i]
```
`break` here (and in the `else` below)? Every call to `Remove` can only remove a single element, right?

marten-seemann commented on allowlist.go:
```diff
@@ -0,0 +1,218 @@
+func (al *Allowlist) Allowed(ma multiaddr.Multiaddr) bool {
+	al.mu.RLock()
+	defer al.mu.RUnlock()
+	ip, err := manet.ToIP(ma)
```
Move mutex lock after this call.

marten-seemann commented on allowlist_test.go:
```diff
@@ -0,0 +1,264 @@
+
+	peerA := test.RandPeerIDFatal(t)
+	peerB := test.RandPeerIDFatal(t)
+	multiaddrA, _ := multiaddr.NewMultiaddr("/ip4/1.2.3.4/tcp/1234")
```
Nit: Use `ma.StringCast`.

MarcoPolo commented:
@marten-seemann anything left?
